mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/events/ElasticSearchListeningMessageSearchIndex.java

```diff
@@ -84,22 +94,63 @@ public void add(MailboxSession session, Mailbox<Id> mailbox, MailboxMessage<Id>
     public void delete(MailboxSession session, Mailbox<Id> mailbox, MessageRange range) throws MailboxException {
         switch (range.getType()) {
             case ALL:
-                indexer.deleteAllMatchingQuery(termQuery(JsonMessageConstants.MAILBOX_ID, mailbox.getMailboxId().serialize()));
+                deleteAllMessages(mailbox);
                 break;
             case FROM:
+                deleteFromMessages(mailbox, range);
+                break;
+            case ONE:
+                deleteSingleMessage(mailbox, range);
+                break;
+            case RANGE:
+                deleteRangeMessage(mailbox, range);
+                break;
+            default:
+                throw new RuntimeException("MessageRange type unknown");
+        }
+    }
+
+    private void deleteRangeMessage(Mailbox<Id> mailbox, MessageRange range) {
+        try {
+            if (range.size() > maximumBulkBatchSize) {
                 indexer.deleteAllMatchingQuery(
                     boolQuery()
                         .must(termQuery(JsonMessageConstants.MAILBOX_ID, mailbox.getMailboxId().serialize()))
                         .must(rangeQuery(JsonMessageConstants.ID).gte(range.getUidFrom())));
-                break;
-            default:
-                range.forEach(messageId -> {
-                    try {
-                        indexer.deleteMessage(indexIdFor(mailbox, messageId));
-                    } catch (Exception e) {
-                        LOGGER.error("Error when deleting index for message " + messageId, e);
-                    }
-                });
+            } else {
+                indexer.deleteMessages(StreamSupport.stream(Spliterators.spliteratorUnknownSize(range.iterator(), Spliterator.ORDERED), true)
```

Comment:
A nicer API from ListeningMessageSearchIndex will remove the need from handling MessageRange as an iterator and will remove this uggly line.